hazizyan/MONIT-24455

hrayrazizyan opened the pull request:
Added the possibility of sending statistics to a different server or customer from the other traces.
Added the possibility to turn on/off statistics reporting through direct ingestion.
Added root names to statistics traces
Added special symbols in front of statistics traces in case of printing to a file.

hrayrazizyan committed:
MONIT-24455:
Added the possibility of sending

hrayrazizyan committed:
MONIT-24455:
Added the possibility of sending statistics to a different server or customer from the other traces.
Added the possibility to turn on/off statistics reporting through direct ingestion.
Added root names to statistics traces
Added special symbols in front of statistics traces in case of printing to a file.

dbaghdasarya commented on applicationConfig.yaml:
```diff
@@ -18,5 +18,12 @@ traceOutputFile: "wftraces.txt"
 ## Uncomment and setup for direct ingestion. The direct ingestion is possible only over https.
-#server: "<cluster>.wavefront.com"
-#token: "bdc66030-a1a8-493b-b416-5559fdcfa45d"
+#server: "meru.wavefront.com"
```
This is a public repo. Please, remove all sensitive data such as server name and token value.

dbaghdasarya commented on src/main/java/com/wavefront/AbstractTraceLoader.java:
```diff
@@ -134,7 +134,9 @@ protected void loadApplicationConfig() throws Exception {
             ApplicationConfig.class);
       }
-      new FileWriter(applicationConfig.getTraceOutputFile(), false).close();
+      if(applicationConfig.getTraceOutputFile() != null){
```
This change is already in `master`. Did you rebase your branch on top of the latest master before pushing it?

dbaghdasarya commented on src/main/java/com/wavefront/SpanSender.java:
```diff
@@ -27,27 +31,35 @@
   private final String traceOutputFile;
   private final DataQueue dataQueue;
   private final AtomicBoolean stopSending = new AtomicBoolean(false);
+  private String reportStat = null;
```
Why `reportStat` is a string?

hrayrazizyan replied:
works fine with boolean

dbaghdasarya commented on src/main/java/com/wavefront/SpanSender.java:
```diff
@@ -77,13 +89,23 @@ public void saveToFile() throws Exception {
-        fileWriter.write("[{\"root\":\"" + tempTrace.getRoot() +
+        StringBuilder stringBuilder = new StringBuilder("[{\"root\":\"" + tempTrace.getRoot() +
             "\"}," + tempTrace.toWFTrace().toJSONString() + "]\n");
+        if(Defaults.STAT_ROOT_NAME_LIST.contains(tempTrace.getRoot())) {
```
As far as I remember this was discussed and considered to check span type instead of root span name.

hrayrazizyan replied:
Works fine with spanKind. Deleting STAT_ROOT_NAME_LIST from Defaults

dbaghdasarya commented on src/main/java/com/wavefront/SpanSender.java:
```diff
@@ -77,13 +89,23 @@ public void saveToFile() throws Exception {
         }
-        roots.add(tempTrace.getRoot());
-        fileWriter.write("[{\"root\":\"" + tempTrace.getRoot() +
+        StringBuilder stringBuilder = new StringBuilder("[{\"root\":\"" + tempTrace.getRoot() +
```
This operation is redundant in case of stat trace and `reportStat=false`.

dbaghdasarya commented on src/main/java/com/wavefront/SpanSender.java:
```diff
@@ -116,7 +138,19 @@ public void run() {
         while (iter.hasNext() && sentSpans < mustBeSentSpans) {
           Span tempSpan = iter.next();
-          if (tempSpan.getStartMillis() < System.currentTimeMillis()) {
+          if(tempSpan.getKind() == SpanKind.STATISTICS && reportStat != null){
```
I guess that here you have to check that `statSender!=null`.

dbaghdasarya commented on src/main/java/com/wavefront/WavefrontTraceLoader.java:
```diff
@@ -63,7 +63,22 @@ void setupSenders() throws IOException {
+            "@" + applicationConfig.getStatServer());
+      }
+      else {
+        statClientFactory.addClient("https://" + applicationConfig.getToken() +
```
What if a user uses `proxy` for ingestion and didn't provide a `serverName` and `token`?
I guess that `statSender` should be `null` in this case.

hrayrazizyan replied:
So if neither server nor statServer have been provided, statistics will not be sent anywhere?

dbaghdasarya replied:
Where to send it? 
If `reportStat=true` you can throw an exception and terminate the App. or just output it in console, like it was done before.

dbaghdasarya commented on src/main/java/com/wavefront/config/ApplicationConfig.java:
```diff
@@ -71,6 +71,22 @@
+   * Token for the statistics server.
+   */
+  @JsonProperty
+  private String reportStat = null;
```
I guess that Boolean is a more suitable type for this parameter.

hrayrazizyan replied:
works fine with boolean

dbaghdasarya commented on src/main/java/com/wavefront/datastructures/Span.java:
```diff
@@ -53,6 +54,23 @@ public Span(String name, long startMillis, long durationMillis, @Nullable String
     this.spanLogs = spanLogs;
```
Code duplication. You can here call the next constructor with `..., SpanKind.REGULAR` value.

dbaghdasarya replied:
I didn't mean to remove the previous constructor. It is a good practice to preserve methods signatures.
I asked you to replace the body with a call of the new constructor with the same parameters and additionally `kind=REGULAR`.

dbaghdasarya commented on src/main/java/com/wavefront/datastructures/StatSpan.java:
```diff
@@ -0,0 +1,20 @@
+import javax.annotation.Nullable;
+
+public class StatSpan extends Span{
+  public StatSpan(String name, long startMillis, long durationMillis, @Nullable String source, UUID traceUUID, UUID spanUUID, @Nullable List<UUID> parents, @Nullable List<UUID> followsFrom, @Nullable List<Pair<String, String>> tags, @Nullable List<SpanLog> spanLogs, SpanKind kind) {
```
I guess that you don't need spanKind parameter for this constructor and can just set `SpanKind.STATISTICS` for the base class.

dbaghdasarya commented on src/main/java/com/wavefront/datastructures/StatSpan.java:
```diff
@@ -0,0 +1,20 @@
+  }
+
+  @Override
+  public String toString() {
```
Do you use this overriding?

hrayrazizyan replied:
for differentiating statistics spans from ordinary ones

dbaghdasarya commented on src/main/java/com/wavefront/generators/TraceGenerator.java:
```diff
@@ -96,6 +104,25 @@ protected void startGeneration(boolean isRealTime, @Nonnull GeneratorConfig gene
+    sendStat(str);
+
+    logger.info("Generation complete!");
+
```
Extra empty lines.

dbaghdasarya commented on src/main/java/com/wavefront/generators/TraceGenerator.java:
```diff
@@ -138,29 +168,23 @@ protected void startGeneration(boolean isRealTime, @Nonnull GeneratorConfig gene
       traceTypeTags.add(new Pair<>("Debug spans count", Integer.toString(v.getDebugSpansCount())));
 
-      Span type_stat = new Span(k,
+      StatSpan type_stat = new StatSpan(k,
```
We use `camelCase` for variables naming.

dbaghdasarya commented on src/main/java/com/wavefront/generators/TraceGenerator.java:
```diff
@@ -138,29 +168,23 @@ protected void startGeneration(boolean isRealTime, @Nonnull GeneratorConfig gene
       statTrace.add(1, type_stat);
+
+      statTrace.setRoot(rootName);
+
```
Extra empty lines.

dbaghdasarya commented on src/main/java/com/wavefront/generators/TraceGenerator.java:
```diff
@@ -3,13 +3,21 @@
+import com.wavefront.datastructures.SpanKind;
+import com.wavefront.datastructures.StatSpan;
 import com.wavefront.datastructures.Trace;
+import com.wavefront.internal.reporter.WavefrontInternalReporter;
```
Do you really need all these new imports?

dbaghdasarya commented on src/main/java/com/wavefront/helpers/Defaults.java:
```diff
@@ -43,5 +43,9 @@
 
   public static final List<ValueDistribution> DEFAULT_SPANS_DISTRIBUTIONS = List.of(
       new ValueDistribution(3, RANDOM.nextInt(10) + 5, 100));
+  public static final List<String> STAT_ROOT_NAME_LIST = List.of(
```
I guess that TraceTypePatternValidator should be updated to prohibit the usage of these words as a service name.

hrayrazizyan committed:
MONIT-24455:
Added the possibility of sending statistics to a different server or customer from the other traces.
Added the possibility to turn on/off statistics reporting through direct ingestion.
Added root names to statistics traces
Added special symbols in front of statistics traces in case of printing to a file.

dbaghdasarya commented on src/main/java/com/wavefront/AbstractTraceLoader.java:
```diff
@@ -135,6 +135,7 @@ protected void loadApplicationConfig() throws Exception {
             ApplicationConfig.class);
       }
       new FileWriter(applicationConfig.getTraceOutputFile(), false).close();
+
```
You haven't actually changed this file. Could you please revert it so it doesn't participate in the review?

dbaghdasarya commented on src/main/java/com/wavefront/WavefrontTraceLoader.java:
```diff
@@ -63,7 +63,27 @@ void setupSenders() throws IOException {
+        if (applicationConfig.getStatServer() != null && applicationConfig.getStatToken() != null) {
+          statClientFactory.addClient("https://" + applicationConfig.getStatToken() +
+              "@" + applicationConfig.getStatServer());
+        } else {
```
This if..if..else combination is complicated for understanding and checking. Please rework it.
BTW, what if `getServer()!=null` but `getToken()==null`?

hrayrazizyan committed:
MONIT-24455:
Added the possibility of sending statistics to a different server or customer from the other traces.
Added the possibility to turn on/off statistics reporting through direct ingestion.
Added root names to statistics traces
Added special symbols in front of statistics traces in case of printing to a file.

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromPatternGenerator.java:
```diff
@@ -134,7 +135,8 @@ protected Trace generateTrace(long startMillis) {
         getTags(traceTypePattern, traceTypePattern.traceTypeName, traceTypePattern.errorRate),
-        null));
+        null,
+        SpanKind.REGULAR));
```
This should be updated when you rework the Span constructor.

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromPatternGenerator.java:
```diff
@@ -166,7 +168,8 @@ protected Trace generateTrace(long startMillis) {
               getTags(traceTypePattern, spanName, 0),
-              null));
+              null,
+              SpanKind.REGULAR));
```
The same is here.

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromTopologyGenerator.java:
```diff
@@ -92,7 +93,8 @@ private void generateTraceTemplates() {
           List.of(new Pair<>(SERVICE, root)),
-          null));
+          null,
+          SpanKind.REGULAR));
```
the same is here.

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromTopologyGenerator.java:
```diff
@@ -127,7 +129,8 @@ private void generateTraceTemplates() {
               List.of(new Pair<>(SERVICE, nextService)),
-              null));
+              null,
+              SpanKind.REGULAR));
```
The same is here.

dbaghdasarya commented on src/main/java/com/wavefront/datastructures/SpanFromWF.java:
```diff
@@ -113,6 +113,6 @@ public Span toSpan() {
 
     return new Span(name, startMs, durationMs, host, UUID.fromString(traceId),
-        UUID.fromString(spanId), parents, followsFrom, tags, null);
+        UUID.fromString(spanId), parents, followsFrom, tags, null, SpanKind.REGULAR);
```
Return the old constructor

hrayrazizyan committed:
MONIT-24455:
Added the possibility of sending statistics to a different server or customer from the other traces.
Added the possibility to turn on/off statistics reporting through direct ingestion.
Added root names to statistics traces
Added special symbols in front of statistics traces in case of printing to a file.

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromPatternGenerator.java:
```diff
@@ -8,6 +8,7 @@
 import com.wavefront.datastructures.RandomDistributionIterator;
 import com.wavefront.datastructures.ReferenceDistribution;
 import com.wavefront.datastructures.Span;
+import com.wavefront.datastructures.SpanKind;
```
Do you need to import SpanKind here?

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromTopologyGenerator.java:
```diff
@@ -9,6 +9,7 @@
 import com.wavefront.datastructures.RandomDistributionIterator;
 import com.wavefront.datastructures.ReferenceDistribution;
 import com.wavefront.datastructures.Span;
+import com.wavefront.datastructures.SpanKind;
```
The same is here. (import SpanKind)

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromTopologyGenerator.java:
```diff
@@ -259,8 +262,8 @@ protected Trace generateTrace(long startMillis) {
-        null
-    ));
+        null,
+        SpanKind.REGULAR));
```
Restore the call of the old constructor.

hrayrazizyan replied:
This fragment is already fixed.

dbaghdasarya commented on src/main/java/com/wavefront/generators/FromTopologyGenerator.java:
```diff
@@ -300,8 +303,8 @@ protected Trace generateTrace(long startMillis) {
-            null
-        ));
+            null,
+            SpanKind.REGULAR));
```
The same is here.

hrayrazizyan replied:
This fragment is already fixed.

hrayrazizyan committed:
Deleted the imports.
